packages/algorand-js/src/runtime/runtime.ts

```diff
@@ -0,0 +1,215 @@
+        const encodedTxnObj = tx.get_obj_for_encoding() as Txn;
+        encodedTxnObj.txID = tx.txID();
+        txns.push(encodedTxnObj);
+      }
+      assignGroupID(txns); // assign unique groupID to all transactions in the array/group
+      return [txns[0], txns]; // by default current txn is the first txn (hence txns[0])
+    } else {
+      // if not array, then create a single transaction
+      const mockParams = mockSuggestedParams(txnParams.payFlags);
+      const tx = mkTransaction(txnParams, mockParams);
+
+      const encodedTxnObj = tx.get_obj_for_encoding() as Txn;
+      encodedTxnObj.txID = tx.txID();
+      return [encodedTxnObj, [encodedTxnObj]];
+    }
+  }
+
+  prepareInitialState (txnParams: ExecParams| ExecParams[], accounts: SDKAccount[]): Context {
+    const [tx, gtxs] = this.createTxnContext(txnParams); // get current txn and txn group (as encoded obj)
+    this.createStatefulContext(accounts); // initialize state before execution
+    return {
+      state: this.store, // state is a copy of store
+      tx: tx,
+      gtxs: gtxs
+    };
+  }
+
+  // updates account balance as per transaction parameters
+  updateBalance (txnParam: ExecParams, account: SDKAccount): void {
+    switch (txnParam.type) {
+      case TransactionType.TransferAlgo: {
+        switch (account.address) {
+          case txnParam.fromAccount.addr: {
+            account.amount -= txnParam.amountMicroAlgos; // remove 'x' algo from sender
+            break;
+          }
+          case txnParam.toAccountAddr: {
+            account.amount += txnParam.amountMicroAlgos; // add 'x' algo to receiver
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Description: update current state and account balances
+   * @param txnParams : Transaction parameters
+   * @param accounts : accounts passed by user
+   */
+  prepareFinalState (txnParams: ExecParams | ExecParams[], accounts: SDKAccount[]): void {
+    if (Array.isArray(txnParams)) { // if txn is a group, update balance as per 'each' transaction
+      for (const txnParam of txnParams) {
+        for (const acc of accounts) {
+          this.updateBalance(txnParam, acc);
+        }
+      }
+    } else {
+      // for a single (stand alone) transaction
+      for (const acc of accounts) {
+        this.updateBalance(txnParams, acc);
+      }
+    }
+  }
+
+  /**
+   * Description: this function executes a transaction based on a smart
+   * contract logic and updates state afterwards
+   * @param txn : Transaction parameters
+   * @param fileName : smart contract file (.teal) name in assets/
+   * @param args : external arguments to smart contract
+   * @param accounts : accounts passed by the user
+   */
+  async executeTx (txnParams: ExecParams | ExecParams[], fileName: string,
+    args: Uint8Array[], accounts: SDKAccount[]): Promise<void> {
+    const context = this.prepareInitialState(txnParams, accounts); // prepare initial state
```

Comment:
let's refactore the functions above, and make more clear function names.

Reply:
yes, i have removed this function.